rust-runtime/aws-smithy-http/src/byte_stream/bytestream_util.rs

```diff
@@ -236,11 +304,316 @@ impl Body for PathBody {
+            .build()
+            .await
+            .is_ok());
+
+        assert!(FsBuilder::new()
+            .path(&file)
+            // Larger than 30 bytes, this will cause an error
+            .length(Length::Exact(31))
+            .build()
+            .await
+            .is_err());
+    }
+
+    #[tokio::test]
+    async fn fsbuilder_supports_offset() {
+        let mut file = NamedTempFile::new().unwrap();
+        let line_0 = "Line 0\n";
+        let line_1 = "Line 1\n";
+
+        write!(file, "{}", line_0).unwrap();
+        write!(file, "{}", line_1).unwrap();
+
+        // Ensure that the file was written to
+        file.flush().expect("flushing is OK");
+
+        let body = FsBuilder::new()
+            .path(&file)
+            // We're going to skip the first line by using offset
+            .offset(line_0.len() as u64)
+            .build()
+            .await
+            .unwrap();
+
+        let data = body.collect().await.unwrap().into_bytes();
+        let data_str = String::from_utf8(data.to_vec()).unwrap();
+
+        assert_eq!(&data_str, line_1);
+    }
+
+    #[tokio::test]
+    async fn fsbuilder_offset_and_length_work_together() {
+        let mut file = NamedTempFile::new().unwrap();
+        let line_0 = "Line 0\n";
+        let line_1 = "Line 1\n";
+        let line_2 = "Line 2\n";
+
+        write!(file, "{}", line_0).unwrap();
+        write!(file, "{}", line_1).unwrap();
+        write!(file, "{}", line_2).unwrap();
+
+        // Ensure that the file was written to
+        file.flush().expect("flushing is OK");
+
+        let body = FsBuilder::new()
+            .path(&file)
+            // We're going to skip line 0 by using offset
+            .offset(line_0.len() as u64)
+            // We want to read only line 1 and stop before we get to line 2
+            .length(Length::Exact(line_1.len() as u64))
+            .build()
+            .await
+            .unwrap();
+
+        let data = body.collect().await.unwrap().into_bytes();
+        let data_str = String::from_utf8(data.to_vec()).unwrap();
+
+        assert_eq!(&data_str, line_1);
+    }
+
+    #[tokio::test]
+    #[should_panic(
+        expected = "offset must be less than or equal to file size but was greater than"
+    )]
+    async fn fsbuilder_with_offset_greater_than_file_length_returns_error() {
+        let mut file = NamedTempFile::new().unwrap();
+        let line_0 = "Line 0\n";
+        let line_1 = "Line 1\n";
+
+        write!(file, "{}", line_0).unwrap();
+        write!(file, "{}", line_1).unwrap();
+
+        // Ensure that the file was written to
+        file.flush().expect("flushing is OK");
+
+        FsBuilder::new()
+            .path(&file)
+            // We're going to skip all file contents by setting an offset
+            // much larger than the file size
+            .offset(9000)
+            .build()
+            .await
+            // this will panic
+            .unwrap();
```

Comment:
There are a few other unwraps in this test, so it could potentially pass even if the code under test is broken. It would be better to do an `assert!(matches!(` with the expected error type here.